lib/lightning/accounts.ex

```diff
@@ -24,33 +45,54 @@ defmodule Lightning.Accounts do
       # coveralls-ignore-stop
     end)
 
-    # Remove credentials
-    Credentials.list_credentials_for_user(id)
-    |> Enum.each(&Credentials.delete_credential/1)
-
-    # Revoke access to projects
+    # Delete all project_users
     Projects.get_projects_for_user(%User{id: id})
     |> Repo.preload(:project_users)
-    |> Enum.each(fn p ->
-      Projects.update_project(
-        p,
-        %{
-          "project_users" => %{
-            "0" => %{
-              "delete" => "true",
-              "user_id" => id,
-              "id" =>
-                Enum.find(p.project_users, fn pu -> pu.user_id == id end)
-                |> Map.get(:id)
-            }
-          }
-        }
-      )
+    |> Enum.map(fn p ->
+      {p, %{"project_users" => project_users_to_params(p.project_users, id)}}
     end)
+    |> Enum.each(fn {project, param} ->
+      Projects.update_project(project, param)
+    end)
+
+    # Prep credentials for deletion
+    credentials_for_user = Credentials.list_credentials_for_user(id)
 
-    User
-    |> Repo.get(id)
-    |> delete_user()
+    user_credentials_ids =
+      Enum.map(credentials_for_user, fn credential -> credential.id end)
+
+    project_credentials_for_user =
+      from(project_credential in Lightning.Projects.ProjectCredential,
+        where: project_credential.credential_id in ^user_credentials_ids
+      )
+      |> Repo.all()
+
+    user_project_credentials_ids =
+      Enum.map(project_credentials_for_user, fn project_credential ->
+        project_credential.id
+      end)
+
+    # Set null all jobs using credentials
+    :ok =
+      from(job in Lightning.Jobs.Job,
+        where: job.project_credential_id in ^user_project_credentials_ids
+      )
+      |> Repo.all()
+      |> Enum.each(fn job ->
+        Lightning.Jobs.update_job(job, %{credential_id: nil})
+      end)
```

Comment:
Again this can be addressed with a query, avoiding loading stuff into memory.

```suggestion
project_credentials_for_user =
  Ecto.assoc(user, [:credentials, :project_credentials])

from(j in Job,
  join: pc in subquery(project_credentials_for_user),
  on: pc.id == j.project_credential_id
)
|> Repo.update_all(set: [project_credential_id: nil])
```

Please look at all the queries here where you are iterating over lists and see how it can be written using queries only.

See: 
https://hexdocs.pm/ecto/Ecto.html#assoc/3
https://hexdocs.pm/ecto/Ecto.Repo.html#c:update_all/3
https://hexdocs.pm/ecto/Ecto.Repo.html#c:delete_all/2

Reply:
Done